docs/design/020-opening-files-in-vu-context.md

```diff
@@ -0,0 +1,101 @@
+
+### Requirements
+
+- **File access in VU stage**: The adopted solution should permit file access both within the VU stage and the init stage.
+- **Idempotency and reproducibility**: The adopted solution should ensure that the archives remain reproducible and idempotent, irrespective of where files are accessed.
+- **Backward compatibility**: The adopted solution should be compatible with the current k6 archive command and should not require any change to the command itself. Namely, it should remain possible to open files in the init stage.
+- **Developer experience**: The adopted solution should improve the developer experience and limit the cost of adopting the new file handling.
+- **Error handling**: The adopted solution should allow for and provide precise error handling and transparent reporting, especially when dealing with file access issues or conflicts in the VU stage (which is a strength of the current way of handling files).
+- **Backward compatibility**: The adopted solution should refrain from introducing breaking changes to the current solution. 
+
+## Possible Solutions
+
+### Solution A: Prior declaration of resources
+
+#### Using a dedicated API
+
+We can provide an API, possibly as part of the new fs module, for users to explicitly include files.
+
+```javascript
+import { open, include } from "k6/experimental/fs";
+
+include('my/source/file.txt')
+include('my/dir')
+
+export default function () {
+    // succeeds
+    await open('my/source/file.txt');
+    
+    // fails with 'resource csvs/data.csv not included by k6, ...'
+    await open('csvs/data.csv')
+}
+```
+
+**Key Points**:
+- The `include` function is only usable in the init stage. Invoking it within VU code will trigger an error.
+- Using `include` ensures a file's inclusion during the archive process, making it accessible in the VU stage.
+- Opening an un-included file in the VU stage will result in an error.
+Files opened in the init stage without prior inclusion should still operate as they currently do (though this is debatable) to maintain backward compatibility.
+
+#### Using dedicated option set
+
+In addition to, or instead of the API, users can specify which files to include via a set of options.
+
+```javascript
+export const options = {
+    // ...
+    
+    include: [
+        'my/source/file.txt',
+        'testdata',
+    ]
+}
+
+export default function () {
+        // succeeds
+    await open('my/source/file.txt');
+    
+    // fails with 'resource csvs/data.csv not included by k6, ...'
+    await open('csvs/data.csv')
+}
+```
+
+**Key Points**:
+- The rules from the explicit API still apply: if a file isn't referenced in the options and is opened within the VU stage, an error occurs.
+- Attempting to open an un-referenced file within the init stage may not result in an error, ensuring backward compatibility.
+
+### Solution B: Awaiting Proposals
+
+Have a novel idea? Please share it with us in a dedicated commit :bow:
```

Comment:
I would argue we do not need a *new* API at all.

There are two problems:
1. you can't *at* all `open` files outside of init code
2. it might be beneficial to have an API that tells k6 to allow a given file to be openable outside of the init code.

There is also the sligth twist that *actually* `open` will return you an error even in init code if you try to open a file in a VU after the initial one that was *not* opened by the initial one.

https://github.com/grafana/k6/pull/2314

```javascript
open("./someotherfile.txt") // will not error out for either the initial nor the VU with id `1`
if (__VU == 1) {
  open("./somefile.txt") // will always error out as the initial VU did not open it.
}
```

Again the first `open` is called both in VU with id `0` but also for any VU that will execute a default function (for example).  See https://github.com/grafana/k6/issues/1771 for the rational and discussion.

the tl;dr is that if we want to allow `open` to at all execute in vu code we can have the same limitation - the file needs to be opened in the VU==0. This is one decision and change.

Adding API/configuration that adds files (again in the initial VU as discussed in #1771) is IMO a separate discussion.